src/pyclass.rs

```diff
@@ -107,120 +107,306 @@ pub trait PyClass:
-    }
-
-    // nb_bool is a part of PyObjectProtocol, but should be placed under tp_as_number
-    let mut nb_bool = None;
-    // basic methods
-    if let Some(basic) = T::basic_methods() {
-        unsafe { basic.as_ref() }.update_typeobj(type_object);
-        nb_bool = unsafe { basic.as_ref() }.nb_bool;
-    }
-
-    // number methods
-    type_object.tp_as_number = T::number_methods()
-        .map(|mut p| {
-            unsafe { p.as_mut() }.nb_bool = nb_bool;
-            p.as_ptr()
-        })
-        .unwrap_or_else(|| nb_bool.map_or_else(ptr::null_mut, ffi::PyNumberMethods::from_nb_bool));
-    // mapping methods
-    type_object.tp_as_mapping = T::mapping_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
-    // sequence methods
-    type_object.tp_as_sequence = T::sequence_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
-    // async methods
-    type_object.tp_as_async = T::async_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
-    // buffer protocol
-    type_object.tp_as_buffer = T::buffer_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
+    let mut slots = vec![];
 
     let (new, call, mut methods) = py_class_method_defs::<T>();
-
+    maybe_push_slot(&mut slots, ffi::Py_tp_new, new.map(|v| v as *mut c_void));
+    maybe_push_slot(&mut slots, ffi::Py_tp_call, call.map(|v| v as *mut c_void));
     // normal methods
     if !methods.is_empty() {
         methods.push(ffi::PyMethodDef_INIT);
-        type_object.tp_methods = Box::into_raw(methods.into_boxed_slice()) as _;
+        maybe_push_slot(
+            &mut slots,
+            ffi::Py_tp_methods,
+            Some(Box::into_raw(methods.into_boxed_slice()) as *mut c_void),
+        );
     }
 
-    // __new__ method
-    type_object.tp_new = new;
-    // __call__ method
-    type_object.tp_call = call;
-
-    // properties
-    let mut props = py_class_properties::<T>();
-
-    if !T::Dict::IS_DUMMY {
-        props.push(ffi::PyGetSetDef_DICT);
+    if let Some(basic) = T::basic_methods() {
+        unsafe { basic.as_ref() }.update_slots(&mut slots);
     }
-    if !props.is_empty() {
-        props.push(ffi::PyGetSetDef_INIT);
-        type_object.tp_getset = Box::into_raw(props.into_boxed_slice()) as _;
+
+    if let Some(number) = T::number_methods() {
+        push_numbers_slots(&mut slots, unsafe { number.as_ref() })
     }
 
-    // set type flags
-    py_class_flags::<T>(type_object);
+    slots.push(ffi::PyType_Slot {
+        slot: 0,
+        pfunc: ptr::null_mut(),
+    });
+    let mut spec = ffi::PyType_Spec {
+        name: match module_name {
+            Some(module_name) => CString::new(format!("{}.{}", module_name, T::NAME))?.into_raw(),
+            None => CString::new(T::NAME)?.into_raw(),
+        },
+        basicsize: std::mem::size_of::<T::Layout>() as c_int,
+        itemsize: 0,
+        flags: 0, // XXXX: FILL ME IN PROPERLY,
+        slots: slots.as_mut_slice().as_mut_ptr(),
```

Comment:
```suggestion
        slots: Box::into_raw(slots.into_boxed_slice()) as _,
```

Reply:
This leaks the memory -- as far as I can tell, `slots` does not need to stay alive longer than the call (it's not documented, but I read the code).

Reply:
> it's not documented, but I read the code

Happy to know that :+1: